ably.ts.d

```diff
@@ -0,0 +1,457 @@
+    authCallback?: (data: String | TokenDetails | TokenRequest) => void;
+    authHeaders?: Array<any>;
+    authMethod?: String;
+    authParams?: Array<any>;
+    authUrl?: String;
+    key?: String;
+    queryTime?: boolean;
+    token?: String | TokenDetails;
+    tokenDetails?: TokenDetails;
+    useTokenAuth?: boolean;
+  }
+
+  interface TokenParams {
+    capability: String;
+    clientId?: String;
+    nonce?: String;
+    timestamp: Number;   // Time?
+    ttl: Number;
+  }
+
+  interface CipherParams {
+    algorithm: String;
+    key: any;
+    keyLength: number;
+    mode: String;
+  }
+
+  interface ErrorInfo {
+    code: number;
+    message: String;
+    statusCode: number;
+  }
+
+  interface ConnectionDetails {
+    clientId: String;
+    connectionKey: String;
+    connectionStateTtl: number;
+    maxFrameSize: number;
+    maxInboundRate: number;
+    maxMessageSize: number;
+    serverId: String;
+    maxIdleInterval: number;
+  }
+
+  interface StatsMessageTypes {
+    all: number,
+    messages: number,
+    presence: number;
+  }
+
+  interface StatsRequestCount {
+    failed: number,
+    refused: number,
+    succeeded: number
+  }
+
+  interface StatsResourceCount {
+    mean: number,
+    min: number,
+    opened: number,
+    peak: number,
+    refused: number
+  }
+
+  interface StatsConnectionTypes {
+    all: number,
+    plain: number,
+    tls: number
+  }
+
+  interface StatsMessageTraffic {
+    all: number,
+    realtime: number,
+    rest: number,
+    webhook: number
+  }
+
+  // Listeners
+  type ListenerFn = (...args: Array<any>) => void;
+
+
+  // Internal Classes
+  declare class EventEmitter {
+    on: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+    once: (eventOrCallback: String | ListenerFn, callback?: ListenerFn) => void;
+
+    off: (eventOrCallback?: String | ListenerFn, callback?: ListenerFn) => void;
+
+    emit: (Event: String, ...args: Array<any>) => void;
+  }
+
+  // Classes
+  export declare class Auth {
+    clientId: String;
+    authorize: (tokenParams?: TokenParams, authOptions?: AuthOptions) => TokenDetails;
+    createTokenRequest: (tokenParams?: TokenParams, authOptions?: AuthOptions) => TokenRequest;
+    requestToken: (TokenParams?: TokenParams, authOptions?: AuthOptions) => TokenDetails;
+  }
+
+  export declare class TokenDetails {
+    fromJson: (JsonData: string) => TokenDetails;
+    capability: String;
+    clientId: String;
+    expires: number; // Time
+    issued: number; // Time
+    token: String;
+  }
+
+  export declare class TokenRequest {
+    fromJson: (JsonData: String) => TokenRequest;
+    capability: String;
+    clientId: String;
+    keyName: String;
+    mac: String;
+    nonce: String;
+    timestamp: number; // Time
+    ttl: number;
+  }
+
+  export declare class ChannelOptions {
+    withCipherKey: (key: any) => ChannelOptions;
+    cipher: (params: CipherParams) => void;
+  }
+
+  export declare class RestPresence {
```

Comment:
Actually they don't have to match, but for "our" sake, they should, so that we don't get confused.    The only thing that has to match is that the callback sends a RestPresence object then TS will use the RestPresence definition.

Is this supposed to be named something else, I just copied the IDL...